### PR TITLE
Run GeneratorFilter modules in an external process controlled by cmsRun

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

ExternalGeneratorFilter is a cmsRun module which can be configured to run a module built from GeneratorFilter<> in a separate process. This separate process is controlled by ExternalGeneratorFilter and cooperatively shares the CPU with cmsRun.

#### PR validation:

Additional unit tests were added to compare the output of Pythia8GeneratorFilter run directly within cmsRun and run via the ExternalGeneratorFilter. With just 1 stream/thread the output is identical. With multiple streams, the GenLumiInfoHeader, edm::HepMCProduct, GenEventInfoProduct and the GenRunInfoProduct are identical. The GenLumiInfoProduct is nearly identical exception the XSec value and errors are different.

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by Dr15Jones** [comment]
For reviewers: the existing `GenLumiInfoProduct::mergeProduct` method does not attempt to combine the values for XSec. Instead, we just get the value from the first `GenLumiInfoProduct`. That seems less than optimal. Is there a reasonable way to combine the XSec values to get a more accurate number?

For reference, XSec generated for Pythia8 comes from here

https://github.com/cms-sw/cmssw/blob/cd9d8d7965adc4c0e9c885e3d64e9c48e9571382/GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc#L626-L630

@smrenna any ideas?

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
@dpiparo FYI

**Comment by Dr15Jones** [comment]
I did some simple speed comparisons. I started from a simple Pythia8 configuration
```python
process.generator = cms.EDFilter("Pythia8GeneratorFilter",
    comEnergy = cms.double(7000.),
    PythiaParameters = cms.PSet(
        pythia8_example02 = cms.vstring('HardQCD:all = on',
                                        'PhaseSpace:pTHatMin = 20.'),
        parameterSets = cms.vstring('pythia8_example02')
    )
```
This runs very fast (100,000 events in about 30 minutes) and has no filtering.

- If I use 1 thread and run the generator directly it takes 1815 seconds.
- If I use 1 thread and use the ExternalGeneratorFilter it takes 3035 seconds.
- If I use 4 threads and the ExternalGeneratorFilter it takes 903 seconds

The upshot is for this case where Pythia8 runs very fast per event, the time it takes to serialize/deserialize to/from the external application is a large overhead. However, it does scale better with more threads.

**Comment by qliphy** [comment]
adding @jordan-martins  this should be good for BPH low filter efficiency requests.

**Comment by qliphy** [comment]
> For reviewers: the existing `GenLumiInfoProduct::mergeProduct` method does not attempt to combine the values for XSec. Instead, we just get the value from the first `GenLumiInfoProduct`. That seems less than optimal. Is there a reasonable way to combine the XSec values to get a more accurate number?
> 
> For reference, XSec generated for Pythia8 comes from here
> 
> https://github.com/cms-sw/cmssw/blob/cd9d8d7965adc4c0e9c885e3d64e9c48e9571382/GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc#L626-L630
> 
> @smrenna any ideas?

You may refer to GenXSecAnalyzer which combines GenLumiInfoProduct to get accurate Xsec.
https://github.com/cms-sw/cmssw/blob/master/GeneratorInterface/Core/plugins/GenXSecAnalyzer.cc

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
>You may refer to GenXSecAnalyzer which combines GenLumiInfoProduct to get accurate Xsec.

I used that calculation to handle XSec improvements in call to GenLumiInfoProduct::mergeProduct.

File: GeneratorInterface/Core/bin/externalGenerator.cc [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+  WorkerMonitorThread monitorThread;
+
+  monitorThread.startThread();
+
+  CMS_SA_ALLOW try {
+    std::string const memoryName(vm[kMemoryNameOpt].as<std::string>());
+    std::string const uniqueID(vm[kUniqueIDOpt].as<std::string>());
+    {
+      //using namespace boost::interprocess;
+      //auto controlNameUnique = unique_name(memoryName, uniqueID);
```
Comment (makortel) [inline GeneratorInterface/Core/bin/externalGenerator.cc:136]: Is this commented code still useful?

Comment (Dr15Jones) [reply GeneratorInterface/Core/bin/externalGenerator.cc:136]: I don't think so

Comment (Dr15Jones) [inline GeneratorInterface/Core/bin/externalGenerator.cc:136]:
```suggestion
```

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
+core

**Comment by makortel** [comment]
How can we eventually deploy the external GeneratorFilter? I'm wondering like what steps are needed from the GEN side, and where exactly we can or have to specify that a generator would be run externally (like does it have to into the generator configuration fragments, or can we have some more general switch e.g. in cmsDriver).

(I know this PR is probably not the best place to discuss if the questions are nontrivial, should we open an issue, or how should we proceed?)

**Comment by Dr15Jones** [comment]
What is the plan for evaluating this? This was placed as a high priority by the L1s.

**Comment by silviodonato** [comment]
> What is the plan for evaluating this? This was placed as a high priority by the L1s.

@alberto-sanchez @agrohsje @efeyazgan @mkirsano @qliphy @SiewYan

**Comment by qliphy** [comment]
+1

File: FWCore/SharedMemory/interface/ReadBuffer.h [hunk 2]
```diff
@@ -26,18 +26,19 @@ This works in conjunction with WriteBuffer.
...
+#include "FWCore/SharedMemory/interface/BufferInfo.h"
 
 // forward declarations
 
 namespace edm::shared_memory {
   class ReadBuffer {
   public:
     /** iUniqueName : must be unique for all processes running on a system.
-        iBufferIndex : is a pointer to a shared_memory address where the same address needs to be shared by ReadBuffer and WriteBuffer.
+        iBufferInfo : is a pointer to a shared_memory address where the same address needs to be shared by ReadBuffer and WriteBuffer.
```
Comment (silviodonato) [inline FWCore/SharedMemory/interface/ReadBuffer.h:37]: Have you kept the same description on purpose?

File: FWCore/SharedMemory/interface/WriteBuffer.h [hunk 3]
```diff
@@ -29,21 +29,22 @@
...
+#include "FWCore/SharedMemory/interface/BufferInfo.h"
 
 // forward declarations
 
 namespace edm::shared_memory {
   class WriteBuffer {
   public:
     /** iUniqueName : must be unique for all processes running on a system.
-        iBufferIndex : is a pointer to a shared_memory address where the same address needs to be shared by ReadBuffer and WriteBuffer.
+        iBufferInfo : is a pointer to a shared_memory address where the same address needs to be shared by ReadBuffer and WriteBuffer.
```
Comment (silviodonato) [inline FWCore/SharedMemory/interface/WriteBuffer.h:40]: same here

**Comment by silviodonato** [comment]
+1
@Dr15Jones please have a look at my (trivial) comments above

File: FWCore/SharedMemory/interface/ReadBuffer.h (hunk 2 above)
Comment (Dr15Jones) [reply FWCore/SharedMemory/interface/ReadBuffer.h:37]: Yes

Comment (Dr15Jones) [reply FWCore/SharedMemory/interface/ReadBuffer.h:37]: The purpose of the variable is unchanged, I just changed the name to better reflect what it was doing.

File: FWCore/SharedMemory/interface/WriteBuffer.h (hunk 3 above)
Comment (Dr15Jones) [reply FWCore/SharedMemory/interface/WriteBuffer.h:40]: That was intentional.

**Comment by agrohsje** [comment]
Hi @Dr15Jones @smrenna @qliphy ,
I thought about this a little bit. Wouldn't it make sense to combine the cross sections taking their relative uncertainties into account. Referring to this comment from Chris: 

> For reviewers: the existing `GenLumiInfoProduct::mergeProduct` method does not attempt to combine the values for XSec. Instead, we just get the value from the first `GenLumiInfoProduct`. That seems less than optimal. Is there a reasonable way to combine the XSec values to get a more accurate number?

**Comment by makortel** [comment]
> I thought about this a little bit. Wouldn't it make sense to combine the cross sections taking their relative uncertainties into account. 

Hi @agrohsje, I believe @Dr15Jones already did that (see below)? Or is the logic in `GenXSecAnalyzer` sub-optimal?

https://github.com/cms-sw/cmssw/pull/29445#issuecomment-612929398
> > You may refer to GenXSecAnalyzer which combines GenLumiInfoProduct to get accurate Xsec.
> 
> I used that calculation to handle XSec improvements in call to GenLumiInfoProduct::mergeProduct.

**Comment by srimanob** [comment]
Should we plan to do validation on high statistics relvals for the next pre-release? Could GEN come with few workflows to use this feature before?

**Comment by agrohsje** [comment]
@makortel no. that is exactly what i had in mind. missed the commit from chris, sorry.

**Comment by jordan-martins** [comment]
I did some preliminary test for the BPH dumby test as Chris used and further pointed me and it looks really nice. Previously, this dumby one had, with whatever number of cores (evtGen does not support MP), something around 35 events/lumi. With these implementations we could reach, with 4 cores, around 196 events/lumi. In terms of time/evt, prevously we had something around 0.3s to 0.6s. Now, my test indicates something around 0.0734s (a huge gain in core efficiency).

**Comment by qliphy** [comment]
> Should we plan to do validation on high statistics relvals for the next pre-release? Could GEN come with few workflows to use this feature before?

We have done some local tests as you can see from above what Jordan reported on a BPH request, and also a MadGraph request:
https://github.com/cms-sw/cmssw/pull/28899#issuecomment-585828744

GEN plan to have a dedicated discussions on May 4th:
https://indico.cern.ch/event/894534/
where we will make further plans on validation, and foresee further improvement with MG support with readonly gridpack option.

**Pull request merged**